### DISCUSSION TITLE
Can a cron induced API route call res.revalidate?

### DISCUSSION THREAD

**Discussion by jschuur** [vercel/next.js#49164] [Help]
I have a pages (not app) based app running 13.3 that uses cached data at build time via getStaticProps and a Vercel cronjob to rebuild the page by calling [/api/update](https://github.com/jschuur/stats.learnbyvideo.dev/blob/main/apps/site/src/pages/api/update.ts) every hour to update cached data. At the end, it calls res.validate('/').

It does not look like this is sufficient to trigger the page to be rebuilt. Does the cron invoked function not have the ability to trigger a revalidation?

If I call a dedicated API route directly in the browser that just performs the same revalidation, that works fine. From the cron description, that'all Vercel cronjob do too: [call the public URL](https://vercel.com/docs/cron-jobs#how-cron-jobs-work).

There is currently a [huge thread](https://github.com/vercel/next.js/discussions/42290) about revalidation bugs, but this seems to be for people using the app directly. My problem still persists if I use the latest canary build from today.

I'm not sure if this is a genuine bug or just something you're not supposed to do. I can work a round this by just making a fetch request to a second API route from my cronjob that does the revalidation.

**Comment by jschuur** [comment]
It's possible that my particular implementation is actually the problem. It's a T3/TRPC app and in getStaticProps it does a Tanstack Query prefetch of data that is then read in by the client.

After a revalidation should have happened, if I visit the site I actually see a brief flash of the new data and then it immediately reverts. It's as if the cron induced revalidation attempt couldn't update the data entirely? Very bizarre.

**Comment by thathurtabit** [comment]
I'm currently seeing the same/similar issue with my app.
I'm also using a `trpc/t3` stack, and have a cron function that:

* lives at `/pages/api/my-cron` (it does not need to know about my `trpc` set up)
* runs every 24 hours.
* It makes some calls to an API, returns some data, updates a `db` entry that my page reads from (via `getStaticProps`) that will then need to be invalidated/revalidated
* On successful db update, finish with:

```
try {
      await response.revalidate("/page-to-revalidate");
      return response.json({ revalidated: true });
    } catch {
      // If there was an error, Next.js will continue
      // to show the last successfully generated page
      return response.status(500).send("Error revalidating");
    }

``` 
Takeaways:

* When I call this serverless function from Postman (with token validation) it successfully does everything and revalidates the page
* When this function is run through a cron schedule, it updates my `db` as requested, but fails to revalidate my page

It seems `await response.revalidate("/page-to-revalidate");` does not work running through a cron scheduled task.

**Reply by thathurtabit** [reply]
I'm going to create another serverless function to handle page revalidation and call it from my cron task via a `fetch` request. I'll report back if that works...

**Reply by thathurtabit** [reply]
Update, creating another serverless function to call from my cron function still does **not** work.
Calling the revalidate function directly/manually is the only way I can revalidate an ISR page.
